docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,533 @@
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+To introduce this change we need to deprecate specifying `Mesh*Route` in `spec.targetRef`.
+
+Should we allow selecting MeshGateway in `spec.targetRef` when `Mesh*Route` is configured in `spec.to[].targetRef`? probably this 
+is obsolete as Route will select gateway, and we want to apply policy on a route. Is there a need to make it more specific than route? 
+
+### Affected policies
+
+At the moment we can only target `Mesh*Routes` in MeshTimeout policy. This MADR is based on MeshTimeout policy.
+Looking at what can be configured at Envoy route we can also apply this to our MeshRetry policy. Rate limit is also configurable on route,
+but we are configuring rate limit on inbound traffic, which does not apply to this MADR. 
+
+#### Producer/consumer model for targeting Mesh*Route in other policies
+
+It is worth to mention that rate limit would not work with our producer/consumer model. In our model consumer always has more priority
+but consumer should not override producer rate limit as owner of the service should know best, the amount of traffic it could handle.
+
+For MeshRetry our producer/consumer makes sense since producer config is a suggestion from owner of the service. If consumer
+wants to override it because they don't want to retry, they just want to fail fast or the want to retry on bigger number of errors
+they can configure it by applying consumer policy.
+
+### Merging and applying configurations
+
+Let's start with an example:
+
+```yaml
+# Producer route in backend-ns namespace
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: route-to-backend
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: "/slow-endpoint"
+      default:
+        backendRefs:
+          - kind: MeshService
+            name: backend
+---
+# Producer MeshTimeout targeting whole MeshService
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: timeout-on-backend-service
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+    default:
+      http:
+        requestTimeout: 2s
+        streamIdleTimeout: 1h
+---
+# MeshTimeout targeting producer route
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: timeout-on-backend-route
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+  - targetRef:
+      kind: MeshHTTPRoute
+      name: route-to-backend
+    default:
+      http:
+        requestTimeout: 10s
+```
+
+All of these policies have `spec.targetRef` `Mesh` and these are producer policies so this configuration will be applied 
```

Comment:
In this specific example the order of to-items in the resulting `MeshTimeout` policy will be decided by the alphabetical order of `metadata.name`. The order is important, because:
```yaml
to:
  - targetRef:
      kind: MeshService
      name: backend
    default:
      http:
        requestTimeout: 2s
        streamIdleTimeout: 1h
  - targetRef:
      kind: MeshHTTPRoute
      name: route-to-backend
    default:
      http:
        requestTimeout: 10s
```
and
```yaml
to:
  - targetRef:
      kind: MeshHTTPRoute
      name: route-to-backend
    default:
      http:
        requestTimeout: 10s
  - targetRef:
      kind: MeshService
      name: backend
    default:
      http:
        requestTimeout: 2s
        streamIdleTimeout: 1h
```
produce different Rules.

Reply:
why does it work like this? Shouldn't the two be the same? they target different resources, so why naming should be more important that resource they select? This is counter intuitive

Reply:
I agree, the issue is covered here https://github.com/kumahq/kuma/issues/9151